### PR TITLE
Add DoneWhenAllIdle option to runner

### PR DESCRIPTION
**Pull request by RobiNino**

### CODE REVIEW

File: parallel/runner.go [hunk 1]
```diff
@@ -145,3 +150,34 @@ func (r *runner) Cancel() {
 func (r *runner) Errors() map[int]error {
 	return r.errors
 }
+
+// Define the work as done when all consumers are idle for idleThresholdSeconds.
```
Comment (eyalbe4) [inline parallel/runner.go:154]: I suggest adding after the above - 
The function will wait until all consumers are idle.

File: parallel/runner.go [hunk 2]
```diff
@@ -145,3 +150,34 @@ func (r *runner) Cancel() {
...
+				break
+			}
+
+			idleTimestamp, err := strconv.ParseInt(r.lastActive[i], 10, 64)
+			if err != nil {
+				return errors.New("unexpected idle timestamp on consumer. err: " + err.Error())
+			}
+
+			idleTime := time.Unix(idleTimestamp, 0)
+			now := time.Now()
```
Comment (eyalbe4) [inline parallel/runner.go]: This variable definition seem redundant, because `time.Now()` is used aonly once.

File: parallel/runner.go [hunk 3]
```diff
@@ -145,3 +150,34 @@ func (r *runner) Cancel() {
 func (r *runner) Errors() map[int]error {
 	return r.errors
 }
+
+// Define the work as done when all consumers are idle for idleThresholdSeconds.
+// Can be run by the producer as a go routine right after starting to produce.
+// CAUTION - Might panic if tasks are taking more than idleThresholdSeconds to be produced.
+func (r *runner) DoneWhenAllIdle(idleThresholdSeconds int) error {
+	for {
+		time.Sleep(time.Duration(idleThresholdSeconds) * time.Second)
```
Comment (eyalbe4) [inline parallel/runner.go:160]: Probably better to move the above sleep line to be at the end of the loop, rathen than at beginning, to allow for a possible situation where no sleep is needed.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

File: parallel/runner.go (hunk 3 above)
Comment (RobiNino) [reply parallel/runner.go:160]: Needed to avoid an instant panic when the producer consumer is initialized.

**Pull request merged**